src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/IProjectHotReloadNotificationService.cs

```diff
@@ -0,0 +1,28 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.HotReload
+{
+    /// <summary>
+    /// Mef Export that can be used to be notified when the hot reload state changes for a project
+    /// </summary>
+    [ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.System, Cardinality = ImportCardinality.ExactlyOne)]
+    public interface IProjectHotReloadNotificationService
+    {
+        /// <summary>
+        /// Subscribe to this event to be notified of changes to a projects hot reload state. 
+        /// </summary>
+        event AsyncEventHandler<bool> HotReloadStateChangedAsync;
+
+        /// <summary>
+        /// The current state of hot reload for the project
+        /// </summary>
+        bool ProjectIsInHotReload{ get; }
+
+        /// <summary>
+        /// Called by project systems when it enters or exits a hot reload session.
+        /// </summary>
+        Task SetHotReloadStateAsync(bool isInHotReload);
```

Comment:
Callers of this hold a lock that excludes certain actions, such as stopping hot reload (afaict). Are the event handlers likely to be long running? Do we need to raise these outside of the lock, and would cancellation be helpful?

Reply:
The notification should be after hot reload starts and after it finishes - at least from the debugger/hot reload button perspective, so shouldn't block those operations. I think there should only be 1 or 2 listeners of this event, and I wouldn't expect handlers to do any long running operations in response to this notification - though there is no way to guarantee that.  I could add cancellation, but it doesn't help much since AsyncEvent.InvokeAsync() doesn't have an overload which takes a cancellation token. I debated making the implementation fire and forget so that the invocation of the handlers does not block the caller, but that runs the risk of events being handled out of sequence unless more synchronization is added.